### ISSUE TITLE
Should nkr::interface::type_i's Is_Any_General<>() work when given tags?

### ISSUE THREAD

**Issue by r-neal-kelly** [r-neal-kelly/nkr#9]
Since I've been writing documentation about what this function actually does I realized that it is essentially the same as this pseduo-code, for type implementations:
```cpp
this_type.Get_Template() == other_type.Get_Template();
```

Given a more specific set of operands:
```cpp
type_t<int>.Get_Template() == type_t<short>.Get_Template() == true;
```

Naturally as the system stands, switching the operands gives the same result for type implementations:
```cpp
type_t<short>.Get_Template() == type_t<int>.Get_Template();
```

However, that doesn't hold true for tag implementations at the moment, which is the concern of this issue:
```cpp
// this_tag.Get_Template() == other_type.Get_Template()
type_tg<>.Get_Template() == type_t<int>.Get_Template() == true;

// this_type.Get_Template() == other_tag.Get_Template()
type_t<int>.Get_Template() == type_tg<>.Get_Template() == false;
```

It's not going to be too difficult to update this and make it symmetrical, we just have to change each type implementation of Is_Any_General for type_i. We probably have to update template_i in a similar fashion, but that can wait until we have this done.

This reason this asymmetry exists is because tr<> actually requires the asymmetry. By the way, this is an excellent real-world example of where designing an API around a specific use-case can actually have unforseen problems. (A lot of people always talk about how APIs should be designed only around use-cases, but there are other considerations besides specific use cases. This is a historically continual problem in the C++ standard library.)

Anyways, tr<> takes advantage of this asymmetry because it disallows tags themselves to pass constraints. If you are designing a function with tr<>, you expect an actual type and not a tag representing a type. Allowing that would make every function and type have to defensively check if the thing is a tag, and it's just a big no-no. However, the most obvious solution would be to just return false whenever a tag is the subject.

I do think this is the appropriate move forward to achieve the symmetry of the equals operator, however another concern is just how inappropriate the C++20 standard allowing reverse operator lookup is and the problems that come from that. I think that at the end of the day, this will not suffer the same problems because I think the problem is syntactically specific to C++ more than it is semantically problematic. There should be no compile time error if a tag is tested against any type of another template in the system, and vice versa.

**Comment by r-neal-kelly** [comment]
By the way, I absolutely want to keep this new functionality away from the identity traits and generic traits themselves, it should be specific to this function acting as meta-operator.

**Comment by r-neal-kelly** [comment]
It should be noted that if ever we do want to test for tags, instead of trying to use tr<> we could have a new function, maybe tg_tr<>, maybe even just tg<>.

**Comment by r-neal-kelly** [comment]
Another very tough question is whether or not we should compile-time error if a tag is in the subjects of a tr expression. On the one hand, it would be nice to explicitly state "hey a tag can't be constrained by this kind of expression", but on the other hand it would be introducing a validation of subjects which would make users of a function/class constraint have to potentially deal with the resultant error message, which might be pretty confusing and out of context. Would it be better to just have it not pass constraints and then get a compile time relating to that, or to the fact that it's a tag?

**Comment by r-neal-kelly** [comment]
So now that we will be updating each Is_Any_General func to test for tags, perhaps it's time to do something about the identical nature of Is_Any_General for both type and tag implementations. I _think_ that if we inherit the type implementation on the tag implementation then we might even be able to get away with just using the same Is_Any_Specific definition also because I think it adapts to the overriding alias. But isn't that really confusing? I dislike that about the language so if we do the inherit pattern we will keep the the two Is_Any_Specific defs. However, I'm wondering if we should have a static trait that handles this concept. It _has_ to be different from the regular _tr concept for sure, but it's worth thinking about before making the big update.

**Comment by r-neal-kelly** [comment]
Fixed in 10ae9045f9d0f4f729668515244bf178706c613e. I should point out that this removes the strict ability of accessing the trait operator by itself for identities and generics through the interface, however I don't know of a use-case for that. If ever we need it, we can simply have another static function just for that.
